src/net/ws_telnet.cc

```diff
@@ -0,0 +1,191 @@
+      lwsl_info("LWS_CALLBACK_ESTABLISHED\n");
+
+      auto port = (port_def_t *)lws_context_user(lws_get_context(wsi));
+      auto fd = lws_get_socket_fd(lws_get_network_wsi(wsi));
+
+      sockaddr_storage addr = {0};
+      socklen_t addrlen = sizeof(addr);
+      auto result = getpeername(fd, reinterpret_cast<sockaddr *>(&addr), &addrlen);
+      if (result) {
+        lwsl_warn("LWS_CALLBACK_ESTABLISHED: getpeername error, %d\n", evutil_socket_geterror(fd));
+        return -1;  // TODO: maybe do something else?
+      }
+
+      // Process X-REAL-IP
+      {
+        char buf[64];  // maximum characters of ip address is 45.
+        auto buflen = lws_hdr_copy(wsi, buf, sizeof(buf), WSI_TOKEN_HTTP_X_REAL_IP);
+        if (buflen > 0) {
+          struct evutil_addrinfo hints = {0};
+          hints.ai_family = AF_UNSPEC;
+          hints.ai_socktype = SOCK_STREAM;
+          hints.ai_flags = AI_NUMERICHOST;
+          hints.ai_protocol = 0; /* Any protocol */
+
+          struct evutil_addrinfo *res = nullptr;
+          auto ret = evutil_getaddrinfo(buf, nullptr, &hints, &res);
+          if (ret) {
+            lwsl_warn("LWS_CALLBACK_ESTABLISHED: invalid X-REAL-IP : %s , error: %s.\n", buf,
+                      evutil_gai_strerror(ret));
+            return false;
+          }
+          if (res && res->ai_addrlen > 0) {
+            memcpy(&addr, res->ai_addr, res->ai_addrlen);
+            addrlen = res->ai_addrlen;
+
+            evutil_freeaddrinfo(res);
+          }
+        }
+      }
+
+      auto ip = new_user(port, fd, reinterpret_cast<sockaddr *>(&addr), addrlen);
+
+      pss->user = ip;
+      pss->buffer = evbuffer_new();
+
+      ip->iflags |= HANDSHAKE_COMPLETE;
+      ip->lws = wsi;
+
+      //handshake complete so lets setup telnet layer
+      ip->telnet = net_telnet_init(ip);
+      send_initial_telnet_negotiations(ip);
+
+      auto base = evconnlistener_get_base(port->ev_conn);
+      event_base_once(
+          base, -1, EV_TIMEOUT,
+          [](evutil_socket_t fd, short what, void *arg) {
+            auto user = reinterpret_cast<interactive_t *>(arg);
+            on_user_logon(user);
+          },
+          (void *)ip, nullptr);
+      break;
+    }
+    case LWS_CALLBACK_CLOSED: {
+      lwsl_info("LWS_CALLBACK_CLOSED: wsi %p\n", wsi);
+
+      auto *ip = pss->user;
+      if (!ip) {
+        return -1;
+      }
+
+      remove_interactive(ip->ob, 0);
+      pss->user = nullptr;
+
+      evbuffer_free(pss->buffer);
+      pss->buffer = nullptr;
+
+      break;
+    }
+    case LWS_CALLBACK_SERVER_WRITEABLE: {
+      lwsl_info("LWS_CALLBACK_SERVER_WRITEABLE\n");
+
+      auto total = evbuffer_get_length(pss->buffer);
+
+      static unsigned char buf[LWS_PRE + 2048];
+      auto numbytes = evbuffer_copyout(pss->buffer, &buf[LWS_PRE], sizeof(buf) - LWS_PRE);
+      if (numbytes > 0) {
+        // TODO: we could use LWS_WRITE_TEXT , however it is much safer to use binary mode, its
```

Comment:
this comment should be removed... as we always going to write in BINARY frame for telnet protocol

Reply:
yeah that was a copy paste thing i probably just left as much of the code as oringal